docs/proposals/cascading-deletion.md

```diff
@@ -0,0 +1,235 @@
+
+* Propagator:
+  * The Propagator is for optimization, not for correctness.
+  * Maintains a DAG of parent-child relations. This DAG stores only name/uid/orphan triplets, not the entire body of every item.
+  * Consists of an *Event Queue* and a single worker.
+  * Watches for create/update/delete events for all resources, enqueues the events to the *Event Queue*.
+  * Worker:
+    * Dequeues an item from the *Event Queue*.
+    * If the item is an creation or update, then updates the DAG accordingly.
+      * If the object has a parent and the parent doesn’t exist in the DAG yet, then apart from adding the object to the DAG, also enqueues the object to the *Dirty Queue*.
+    * If the item is a deletion, then removes the object from the DAG, and enqueues all its children to the *Dirty Queue*.
+  * The propagator shouldn't need to do any RPCs, so a single worker should be sufficient. This makes locking easier.
+  * With the Propagator, we *only* need to run the Scanner when starting the Propagator to populate the DAG and the *Dirty Queue*.
+
+## Changes to existing components
+
+* Storage: we should add a REST storage for Tombstones. The index should be UID rather than namespace/name.
+
+* API Server: when handling a deletion request, if DeleteOptions.OrphanChildren is true, then the API Server either creates a tombstone with TTL if the tombstone doesn't exist yet, or updates the TTL of the existing tombstone. The API Server deletes the object after the tombstone is created.
+
+* Controllers: when creating child objects, controllers need to fill up the their ObjectMeta.ParentReferences field. Objects that don’t have a parent should have the namespace object as the parent.
+
+# Design 2. Finalizer framework + GC
+
+There are two parts in this design. 2A is about the finalizer framework; 2B is about how to use the framework to enforce the orphaning policy.
+
+## 2A. Finalizer framework
+
+## API changes
+
+```
+type ObjectMeta struct {
+	…
+	Finalizers map[string]string
+}
+```
+
+**ObjectMeta.Finalizers**: the map key is an ad hoc, explanatory name(e.g., "orphan"), the value is an identifier for the responsible component that will remove the finalizer from the map. If the ObjectMeta.DeletionTimestamp is non-nil, entry in the map can only be removed.
+
+## New components
+
+* Finalizer:
+  * Like a controller, a finalizer is always running.
+  * A third party can develop their own finalizer. A finalizer doesn't need to be registered with the API server.
+  * Watches for update events that meet two conditions:
+    1. the updated objects has the identifier of the finalizer in ObjectMeta.Finalizers;
+    2. ObjectMeta.DeletionTimestamp is updated from nil to non-nil.
+  * Applies the finalizing logic to the object in the update event.
+  * Removes itself from ObjectMeta.Finalizers after the finalizing logic is completed.
+  * The API server deletes the object after the last finalizer removes itself from the ObjectMeta.Finalizers field.
```

Comment:
I expected something would send a final delete when it observed no more finalizers and not the API server itself.